File: kid-ios/MyWorld/Views/LoginView.swift
import SwiftUI

struct LoginView: View {
    @Environment(AuthManager.self) private var auth

    @State private var email = ""
    @State private var password = ""
    @State private var submitting = false
    @State private var resetMsg: String?
    @State private var resetBusy = false

    var body: some View {
        VStack(spacing: 24) {
            Spacer()
            // The real app icon as the wordmark, with the rounded-square look.
            Image("MyWorldLogo")
                .resizable()
                .scaledToFit()
                .frame(width: 116, height: 116)
                .clipShape(RoundedRectangle(cornerRadius: 26, style: .continuous))
                .shadow(color: .black.opacity(0.12), radius: 10, y: 4)
                .accessibilityLabel("My World")
            VStack(spacing: 2) {
                Text("My World")
                    .font(.system(size: 44, weight: .bold, design: .rounded))
                    .foregroundStyle(Color(hex: "#ff1493"))
                Text("Tap to Talk")
                    .font(.system(size: 18, weight: .semibold, design: .rounded))
                    .foregroundStyle(Color(hex: "#ad1457").opacity(0.8))
            }
            Text("Sign in to your child's board")
                .font(.title3)
                .foregroundStyle(Color(hex: "#6b7280"))

            VStack(spacing: 14) {
                TextField("Email", text: $email)
                    .textContentType(.emailAddress)
                    .keyboardType(.emailAddress)
                    .textInputAutocapitalization(.never)
                    .autocorrectionDisabled()
                    .padding(14)
                    .background(.thinMaterial)
                    .clipShape(RoundedRectangle(cornerRadius: 14))

                SecureField("Password", text: $password)
                    .textContentType(.password)
                    .padding(14)
                    .background(.thinMaterial)
                    .clipShape(RoundedRectangle(cornerRadius: 14))
            }
            .frame(maxWidth: 420)

            if let err = auth.lastError {
                Text(err)
                    .foregroundStyle(.red)
                    .font(.callout)
                    .multilineTextAlignment(.center)
                    .padding(.horizontal)
            }

            Button {
                Task {
                    submitting = true
                    await auth.signIn(email: email, password: password)
                    submitting = false
                }
            } label: {
                Text(submitting ? "Signing in…" : "Sign in")
                    .font(.title3.weight(.semibold))
                    .padding(.vertical, 14)
                    .frame(maxWidth: 420)
                    .foregroundStyle(.white)
                    .background(Color(hex: "#ff1493"))
                    .clipShape(Capsule())
            }
            .disabled(submitting || email.isEmpty || password.isEmpty)

            // Same reset flow as the web login: enter your email, get a link.
            Button {
                Task { await requestReset() }
            } label: {
                Text(resetBusy ? "Sending…" : "Forgot password?")
                    .font(.subheadline.weight(.semibold))
                    .foregroundStyle(Color(hex: "#ad1457"))
            }
            .disabled(resetBusy)
            if let m = resetMsg {
                Text(m)
                    .font(.footnote)
                    .foregroundStyle(Color(hex: "#047857"))
                    .multilineTextAlignment(.center)
                    .padding(.horizontal)
            }
            Spacer()
            Spacer()
        }
        .padding()
        .background(Color(hex: "#fff7fb"))
    }

    private func requestReset() async {
        let addr = email.trimmingCharacters(in: .whitespaces)
        guard addr.contains("@") else {
            resetMsg = "Enter your account email above first, then tap again."
            return
        }
        resetBusy = true
        defer { resetBusy = false }
        do {
            let body = try JSONSerialization.data(withJSONObject: ["email": addr])
            _ = try await APIClient().request(method: "POST", path: "/api/auth/reset-request",
                                              body: body, contentType: "application/json")
            resetMsg = "If that email has an account, a reset link is on its way. Check your inbox."
        } catch {
            resetMsg = "Couldn't send the link — check your connection and try again."
        }
    }
}

File: android-native/app/src/main/java/io/andrewpeterson/myworld/ui/LoginView.kt
package io.andrewpeterson.myworld.ui

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.widthIn
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.Button
import androidx.compose.material3.ButtonDefaults
import androidx.compose.material3.OutlinedTextField
import androidx.compose.material3.OutlinedTextFieldDefaults
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.input.KeyboardType
import androidx.compose.ui.text.input.PasswordVisualTransformation
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import io.andrewpeterson.myworld.LocalAppContainer
import io.andrewpeterson.myworld.net.resetRequest
import io.andrewpeterson.myworld.ui.theme.Brand
import io.andrewpeterson.myworld.ui.theme.hexColor
import kotlinx.coroutines.launch

/** Port of kid-ios `Views/LoginView.swift` — email/password sign-in. */
@Composable
fun LoginView() {
    val c = LocalAppContainer.current
    val scope = rememberCoroutineScope()
    val lastError by c.auth.lastError.collectAsState()

    var email by remember { mutableStateOf(c.auth.lastEmail()) }
    var password by remember { mutableStateOf("") }
    var submitting by remember { mutableStateOf(false) }

    Column(
        Modifier
            .fillMaxSize()
            .background(hexColor("#fff7fb"))
            .verticalScroll(rememberScrollState())
            .padding(24.dp),
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.Center,
    ) {
        Text("🌍", fontSize = 72.sp)
        Spacer(Modifier.height(12.dp))
        Text("My World", fontSize = 44.sp, fontWeight = FontWeight.Bold, color = Brand.pink)
        Text("Tap to Talk", fontSize = 18.sp, fontWeight = FontWeight.SemiBold,
            color = Brand.pinkDeep.copy(alpha = 0.8f))
        Spacer(Modifier.height(8.dp))
        Text("Sign in to your child's board", fontSize = 18.sp, color = Brand.muted)
        Spacer(Modifier.height(24.dp))

        OutlinedTextField(
            value = email, onValueChange = { email = it.trim() },
            label = { Text("Email") },
            singleLine = true,
            keyboardOptions = KeyboardOptions(keyboardType = KeyboardType.Email),
            shape = RoundedCornerShape(14.dp),
            colors = OutlinedTextFieldDefaults.colors(
                focusedBorderColor = Brand.pink, unfocusedBorderColor = Brand.line),
            modifier = Modifier.widthIn(max = 420.dp).fillMaxWidth(),
        )
        Spacer(Modifier.height(14.dp))
        OutlinedTextField(
            value = password, onValueChange = { password = it },
            label = { Text("Password") },
            singleLine = true,
            visualTransformation = PasswordVisualTransformation(),
            keyboardOptions = KeyboardOptions(keyboardType = KeyboardType.Password),
            shape = RoundedCornerShape(14.dp),
            colors = OutlinedTextFieldDefaults.colors(
                focusedBorderColor = Brand.pink, unfocusedBorderColor = Brand.line),
            modifier = Modifier.widthIn(max = 420.dp).fillMaxWidth(),
        )

        lastError?.let {
            Spacer(Modifier.height(12.dp))
            Text(it, color = Color(0xFFDC2626), fontSize = 14.sp, textAlign = TextAlign.Center)
        }

        Spacer(Modifier.height(20.dp))
        Button(
            onClick = {
                scope.launch {
                    submitting = true
                    c.auth.noteEmail(email)
                    c.auth.signIn(email, password)
                    submitting = false
                }
            },
            enabled = !submitting && email.isNotBlank() && password.isNotBlank(),
            shape = CircleShape,
            colors = ButtonDefaults.buttonColors(containerColor = Brand.pink),
            modifier = Modifier.widthIn(max = 420.dp).fillMaxWidth().height(52.dp),
        ) {
            Text(if (submitting) "Signing in…" else "Sign in",
                fontSize = 18.sp, fontWeight = FontWeight.SemiBold)
        }

        // Same reset flow as the web login: enter your email, get a link.
        Spacer(Modifier.height(10.dp))
        var resetMsg by remember { mutableStateOf<String?>(null) }
        var resetBusy by remember { mutableStateOf(false) }
        TextButton(onClick = {
            val addr = email.trim()
            if (!addr.contains("@")) { resetMsg = "Enter your account email above first, then tap again."; return@TextButton }
            scope.launch {
                resetBusy = true
                resetMsg = try {
                    c.api.resetRequest(addr)
                    "If that email has an account, a reset link is on its way. Check your inbox."
                } catch (_: Exception) { "Couldn't send the link. Check your connection and try again." }
                resetBusy = false
            }
        }, enabled = !resetBusy) {
            Text(if (resetBusy) "Sending…" else "Forgot password?",
                fontSize = 14.sp, fontWeight = FontWeight.SemiBold, color = Brand.pinkDeep)
        }
        resetMsg?.let {
            Text(it, fontSize = 13.sp, color = Color(0xFF047857), textAlign = TextAlign.Center,
                modifier = Modifier.widthIn(max = 420.dp))
        }
    }
}
